pkg/kubelet/dockershim/exec.go

```diff
@@ -110,29 +114,48 @@ func (*NativeExecHandler) ExecInContainer(client libdocker.Interface, container
 		return err
 	}
 
+	// if ExecProbeTimeout feature gate is disabled, preserve existing behavior to ignore exec timeouts
+	var execTimeout <-chan time.Time
+	if timeout > 0 && utilfeature.DefaultFeatureGate.Enabled(features.ExecProbeTimeout) {
+		execTimeout = time.After(timeout)
+	} else {
+		// skip exec timeout if provided timeout is 0
+		execTimeout = nil
+	}
+
 	ticker := time.NewTicker(2 * time.Second)
 	defer ticker.Stop()
 	count := 0
 	for {
-		inspect, err2 := client.InspectExec(execObj.ID)
-		if err2 != nil {
-			return err2
-		}
-		if !inspect.Running {
-			if inspect.ExitCode != 0 {
-				err = &dockerExitError{inspect}
+		select {
+		case <-execTimeout:
+			return exec.NewTimeoutError(fmt.Errorf("command %q timed out", strings.Join(cmd, " ")), timeout)
+		// need to use "default" here instead of <-ticker.C, otherwise we delay the initial InspectExec by 2 seconds.
+		default:
+			inspect, inspectErr := client.InspectExec(execObj.ID)
+			if inspectErr != nil {
+				return inspectErr
 			}
-			break
-		}
 
-		count++
-		if count == 5 {
-			klog.Errorf("Exec session %s in container %s terminated but process still running!", execObj.ID, container.ID)
-			break
-		}
+			if !inspect.Running {
+				if inspect.ExitCode != 0 {
+					return &dockerExitError{inspect}
+				}
 
-		<-ticker.C
-	}
+				return nil
+			}
+
+			// Only limit the amount of InspectExec calls if the exec timeout was not set.
+			// When a timeout is not set, we stop polling the exec session after 5 attempts and allow the process to continue running.
+			if execTimeout == nil {
+				count++
+				if count == 5 {
+					klog.Errorf("Exec session %s in container %s terminated but process still running!", execObj.ID, container.ID)
+					return nil
+				}
+			}
```

Comment:
Doesn't this mean that any probe, **without timeout**, taking longer than 10 seconds will automatically be **no** error, independent of what the actual exit of the command is?

Reply:
There have been a number of follow-ups to this PR since it landed so I suggest you take a look at what's committed on the master branch.

Reply:
Ignore my question - interpretation error from my side, on master (and probably on this commit as well) `client.StartExec()` is blocking and thus this code block happens **after** the command has finished in some form. It is 10 seconds maximum for gathering the exit info, although the code on master is slightly different with the same effect.